client-js/main/client/client-factory.js

```diff
@@ -0,0 +1,233 @@
+ *
+ * Ensures that the `ClientOptions` contain list of the `index.js` files generated by
+ * {@link https://github.com/SpineEventEngine/base/tree/master/tools/proto-js-plugin the Protobuf plugin for JS}
+ * and performs registration of types and parsers containing in these files.
+ *
+ * Creation of the concrete implementation of `Client` instances is delegated to inheritors.
+ */
+class AbstractClientFactory {
+
+  /**
+   * Creates a new instance of `Client` implementation in accordance with given options.
+   *
+   * @param {!ClientOptions} options client initialization options
+   * @return {Client} a `Client` instance
+   */
+  static createClient(options) {
+    this._ensureOptionsSufficient(options);
+    this._registerTypes(...options.protoIndexFiles);
+    return this._clientFor(options);
+  }
+
+  /**
+   * Creates a new instance of `Client` implementation in accordance with given options.
+   *
+   * @param {!ClientOptions} options
+   * @return {Client}
+   * @protected
+   */
+  static _clientFor(options) {
+    throw new Error('Not implemented in abstract base')
+  }
+
+  /**
+   * Ensures whether options object is sufficient for client initialization.
+   *
+   * @param {!ClientOptions} options
+   * @protected
+   */
+  static _ensureOptionsSufficient(options) {
+    if (!options) {
+      throw new Error('Unable to initialize client. The `ClientOptions` is undefined.');
+    }
+
+    const indexFiles = options.protoIndexFiles;
+    if (!Array.isArray(indexFiles) || indexFiles.length === 0) {
+      throw new Error('Only a non-empty array is allowed as ClientOptions.protoIndexFiles parameter.');
+    }
+
+    indexFiles.forEach(indexFile => {
+      if (typeof indexFile !== 'object'
+          || !(indexFile.types instanceof Map)
+          || !(indexFile.parsers instanceof Map) ) {
+        throw new Error('Unable to register Protobuf index files.' +
+          ' Check the `ClientOptions.protoIndexFiles` contains files' +
+          ' generated with "io.spine.tools:spine-proto-js-plugin".');
+      }
+    });
+  }
+
+  /**
+   * Registers all Protobuf types provided by the specified modules.
+   *
+   * After the registration, the types can be used and parsed correctly.
+   *
+   * @param protoIndexFiles the index.js files generated by
+   * {@link https://github.com/SpineEventEngine/base/tree/master/tools/proto-js-plugin the Protobuf plugin for JS}
+   * @private
+   */
+  static _registerTypes(...protoIndexFiles) {
+    for (let indexFile of protoIndexFiles) {
+      for (let [typeUrl, type] of indexFile.types) {
+        KnownTypes.register(type, typeUrl);
+      }
+      for (let [typeUrl, parserType] of indexFile.parsers) {
+        TypeParsers.register(new parserType(), typeUrl);
+      }
+    }
+  }
+}
+
+/**
+ * An implementation of the `AbstractClientFactory` that creates instances of `FirebaseClient`.
+ */
+class FirebaseClientFactory extends AbstractClientFactory {
```

Comment:
To decouple things let's move this implementation into a separate file.

Ideally, everything related to Firebase should be put into a single file.